### PR TITLE
Danial/add xapi transformers

### PR DESCRIPTION
**Pull request by danialmalik**
**Description:** 
Add xapi transformers for events

**Configs for xAPI transformation and routing**

```
EVENT_TRACKING_BACKENDS = {
    ...,
    'xapi': {
    'ENGINE': 'eventtracking.backends.async_routing.AsyncRoutingBackend',
    'OPTIONS': {
        'backend_name': 'xapi',
        'processors': [
                {
                    'ENGINE': 'eventtracking.processors.regex_filter.RegexFilter',
                    'OPTIONS': {
                        'filter_type': 'allowlist',
                        'regular_expressions': [
                            'edx.course.enrollment.activated',
                            'edx.course.enrollment.deactivated',
                            'edx.ui.lms.link_clicked',
                            'edx.ui.lms.sequence.outline.selected',
                            'edx.ui.lms.outline.selected',
                            'edx.ui.lms.sequence.next_selected',
                            'edx.ui.lms.sequence.previous_selected',
                            'edx.ui.lms.sequence.tab_selected',
                            'showanswer',
                            'edx.problem.hint.demandhint_displayed',
                            'edx.grades.problem.submitted',
                            'problem_check',
                            'load_video',
                            'edx.video.loaded',
                            'play_video',
                            'edx.video.played',
                            'stop_video',
                            'edx.video.stopped',
                            'complete_video',
                            'edx.video.completed',
                            'pause_video',
                            'edx.video.paused',
                            'seek_video',
                            'edx.video.position.changed',
                        ]
                    }
                },
        ],
        'backends': {
            'caliper': {
                'ENGINE': 'event_routing_backends.backends.events_router.EventsRouter',
                'OPTIONS': {
                    'processors': [
                        {
                            'ENGINE': 'event_routing_backends.processors.xapi.transformer_processor.XApiProcessor',
                            'OPTIONS': {}
                        }
                    ],
                    'backend_name': 'xapi',
                }
            }
        },
    },
}
}
```

**JIRA:** N/A

**Dependencies:** N/A

**Merge deadline:** List merge deadline (if any)

**Installation instructions:** List any non-trivial installation 
instructions.

**Testing instructions:**

.N/A

**Reviewers:**
- [ ] tag reviewer 
- [ ] tag reviewer 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @danialmalik! I've created [OSPR-5105](https://openedx.atlassian.net/browse/OSPR-5105) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: event_routing_backends/helpers.py
Comment (feanil) [inline event_routing_backends/helpers.py:14]: This feels like bad coupling and I'm talking to the TNL team to figure out if there may be an alternative to doing this.

File: event_routing_backends/processors/xapi/event_transformers/enrollment_events.py
Comment (feanil) [inline event_routing_backends/processors/xapi/event_transformers/enrollment_events.py:41]: Why do we need the full course object here?  Can you tell me more about where/how it gets used?

File: event_routing_backends/helpers.py
Comment (feanil) [reply event_routing_backends/helpers.py:14]: It looks like the function that uses this only gets used once, why do we need it?

File: event_routing_backends/helpers.py
Comment (feanil) [inline event_routing_backends/helpers.py:16]: This should be pulled from `student.api` instead, and be added to that apps `api.py` instead of pulling it directly from the model to reduce coupling.

File: event_routing_backends/helpers.py
Comment (feanil) [inline event_routing_backends/helpers.py:66]: I don't think this indirection for a 1 line function is necessary.

I disagree with this but not enough to block merging the change.

File: event_routing_backends/processors/xapi/constants.py
Comment (feanil) [inline event_routing_backends/processors/xapi/constants.py:8]: Why are these commented out?

**Review by feanil — COMMENTED** [review]
@danialmalik I'm still reviewing but here is the feedback I have so far.

**Comment by natabene** [comment]
@Thank you for your contribution.

File: requirements/base.in
Comment (feanil) [inline requirements/base.in:13]: Since this will be running in python 3.8 and later mostly, we can drop `edx-tincan-py35` support and simply use the upstream tincan library.

**Review by feanil — COMMENTED** [review]
Finished the rest of the review.  I think once we get the questions from the last review figuerd out, this will be good to go.

File: requirements/base.in
Comment (sheralim012) [reply requirements/base.in]: done

File: event_routing_backends/processors/xapi/constants.py
Comment (sheralim012) [reply event_routing_backends/processors/xapi/constants.py]: removed these as they were not used anywhere

File: event_routing_backends/backends/events_router.py
Comment (sheralim012) [reply event_routing_backends/backends/events_router.py]: done

File: event_routing_backends/apps.py
Comment (sheralim012) [reply event_routing_backends/apps.py]: done

File: event_routing_backends/processors/xapi/event_transformers/enrollment_events.py
Comment (sheralim012) [reply event_routing_backends/processors/xapi/event_transformers/enrollment_events.py:33]: here we are extracting the course's display name in which the user is being enrolled

File: event_routing_backends/helpers.py
Comment (sheralim012) [reply event_routing_backends/helpers.py:18]: as `anonymous_id_for_user` belongs to the edx-platform repo, are we going to perform changes there alongside this PR?

File: event_routing_backends/processors/xapi/event_transformers/enrollment_events.py
Comment (feanil) [reply event_routing_backends/processors/xapi/event_transformers/enrollment_events.py:32]: This could be done using the course overview object instead which should be a lot faster and cheaper. https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/course_overviews/api.py#L12

**Review by feanil — COMMENTED** [review]
I still need to review the `external_user_id` code  a bit closer but here are some thoguhts  so far.

File: event_routing_backends/helpers.py
Comment (sheralim012) [reply event_routing_backends/helpers.py:43]: according to the test, this is the case of anonymous user clicking course navigation links, but I'm not able to replicate it as course navigation is not visible to an anonymous user

File: event_routing_backends/processors/xapi/transformer.py
Comment (sheralim012) [inline event_routing_backends/processors/xapi/transformer.py:52]: @feanil I was validating generated statements and found that openid is a URI that uniquely identifies the Agent(user) and here we are using UUID. Is there a way to generate a unique URI for user identification?

File: event_routing_backends/helpers.py
Comment (feanil) [reply event_routing_backends/helpers.py:43]: Will you add a comment to that effect here.  I think there are cases where we do have anonymous users with courseware access under certain configurations.

File: event_routing_backends/processors/xapi/transformer.py
Comment (feanil) [reply event_routing_backends/processors/xapi/transformer.py:52]: Let's use `https://openedx.org/users/user-v1:<uuid>` instead.  I believe the existing xAPI integration in enterprise is doing something like this already.

File: event_routing_backends/helpers.py
Comment (feanil) [reply event_routing_backends/helpers.py]: Yes, we should update edx-platform first.  If you open up a PR over there I can review and help land it.

File: event_routing_backends/helpers.py
Comment (sheralim012) [reply event_routing_backends/helpers.py:43]: I updated the docstring to specify this case

File: event_routing_backends/processors/xapi/transformer.py
Comment (sheralim012) [reply event_routing_backends/processors/xapi/transformer.py]: done

**Review by jinder1s — APPROVED** [review]
This looks good and all of Feanil's comment seem to have been addressed. I plan on merging this once I fix the merge conflict, and incrementing version number

**Comment by jinder1s** [comment]
@danialmalik Hi, it looks like this work was done in a fork and I do not have write access to fork. Could you fix the merge conflict. Once you do, I can merge this PR. Thank You

**Comment by edx-webhook** [comment]
@danialmalik 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/event-routing-backends&entry.867055334=https://github.com/edx/event-routing-backends/pull/11&entry.1484655318=https://github.com/danialmalik&entry.752974735=2020-11-04+12:08&entry.1917517419=2020-12-10+15:57&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
